pkg/cafs/reader.go

```diff
@@ -201,115 +273,332 @@ func (r *chunkReader) WriteTo(writer io.Writer) (n int64, err error) {
 
 }
 
-func calculateKeyAndOffset(off int64, leafSize uint32) (index int64, offset int64) {
-	index = off / int64(leafSize)
+func (r *chunkReader) watchPrefetched(wg *sync.WaitGroup, fetchC <-chan fetch) {
+	defer wg.Done()
+
+	// collect prefetched leaf index
+	for f := range fetchC {
+		r.fetchingLatch.Lock()
+		r.fetching[f.index] = f
+		r.fetchingLatch.Unlock()
+	}
+}
+
+func (r *chunkReader) doPrefetch(index, initiator int, logger *zap.Logger) (LeafBuffer, bool, error) {
+	// doPrefetch is in charge of launching go routines to fetch-ahead of the current leaf whenever appropriate:
+	// * fetching ahead is throttled by some parameter (default is 1 ahead)
+	// * we request leaves ahead only once, unless they have been wiped from buffer cache (reduce wasted work)
+	// * we cannot totally rule out wasted work because the buffer cache may not be able to retain all buffers
+	//
+	// A way to assert that there is no more wasted work than allowed by memory constraint:
+	// $ DEBUG_TEST=1 go test -v -run Reader_All|grep -E '(wasted work)|(has read leaf)'|jq -c '{key: .key,msg: .msg}'|sort
+
+	r.fetchingLatch.Lock()
+	defer r.fetchingLatch.Unlock()
+
+	if _, beingFetched := r.fetching[index+1]; !beingFetched {
+		// starts immediately fetching next blob(s) in the background
+		isRunning := r.seekAhead(index, initiator)
+		if isRunning {
+			// launched or about to be
+			r.fetching[index+1] = fetch{}
+		}
+	}
+
+	if f, beingFetched := r.fetching[index]; beingFetched {
+		// the current leaf is being prefetched
+		if index != initiator {
+			// skip redundant prefetching
+			return nil, false, nil
+		}
+		if f.err != nil {
+			return nil, false, f.err
+		}
+		delete(r.fetching, index)
+
+		r.lruLatch.Lock()
+		b, ok := r.lru.Get(r.pather(r.keys[index]))
+		if ok {
+			lb := b.(LeafBuffer)
+			lb.Pin()
+			r.lruLatch.Unlock()
+			// happy path: prefetching work is reused
+			return lb, true, nil
+		}
+		r.lruLatch.Unlock()
+		// previous work wasted: prefetch has completed but cache entry has been evicted: must restart
+		logger.Debug("restart fetching wasted work")
+		atomic.AddUint64(&r.wasted, 1)
+	}
+
+	// the current leaf is not being prefetched: carry on with read ops.
+
+	// acquire a new buffer from the freelist: the buffer will be kept busy unless an error occurs
+	lb := r.leafPool.Get()
+	lb.Pin()
+	r.fetching[index] = fetch{}
+	return lb, false, nil
+}
+
+func calculateKeyAndOffset(off int64, leafSize uint32) (index int, offset int64) {
+	index = int(off / int64(leafSize))
 	offset = off % int64(leafSize)
 	return
 }
 
-func (r *chunkReader) ReadAt(p []byte, off int64) (totread int, err error) {
+func readLeafFunc(r *chunkReader) func(Key, int, int, <-chan struct{}) (LeafBuffer, bool, error) {
```

Comment:
This is a busy function, please add a comment to the signature on what this function will do. It also triggers the prefetch mechanism in addition to reading the leaf.

Reply:
yeah as it used to... Ok to add comments.